drivers: sensor: ina23x: add support for INA237

bbilas opened the pull request:
Extend the driver of support for INA237 variant which is quite similar to the already supported INA230. Trigger mode support is currently not supported. Changes tested using `sensor-flash`

```
uart:~$ sensor get INA237
channel idx=30 voltage =  11.750000
channel idx=31 current =   1.069900
channel idx=32 power =  12.570100
```

Signed-off-by: Bartosz Bilas <b.bilas@grinn-global.com>

gmarull reviewed (commented):
Quick question: is it feasible to have both variants mounted in the same board? With these changes only one can be used at the same time

bbilas commented:
@gmarull heh you got me ;) I haven't thought about that at all and it seems that it has to be more flexible to handle more than one sensor.

bbilas commented:
I was trying to add some runtime mechanism to allow detecting which variant is being used and based on that, use different registers map but it was starting to look a bit complicated so I'm thinking about splitting this driver. One for 230/231 and one for 237. What do you think?

gmarull commented:
> I was trying to add some runtime mechanism to allow detecting which variant is being used and based on that, use different registers map but it was starting to look a bit complicated so I'm thinking about splitting this driver. One for 230/231 and one for 237. What do you think?

I guess that you can create a separate driver, but re-using as much as possible (shared codebase between 230/1 and 237), right? I'm not familiar enough here to suggest an option :-)

bbilas commented:
Here https://github.com/bbilas/zephyr/commit/a0328d05aa95391e059c8593c04a7eeb3b334fb0 is a proposal with the runtime detection of the INA variant but I would rather create a separate driver for 237 and share the codebase between them.

bbilas committed:
drivers: sensor: ina23x: add ina23x common module

That's needed to not duplicate code between INA230/231
and the upcoming INA237 variant.

Signed-off-by: Bartosz Bilas <b.bilas@grinn-global.com>

bbilas committed:
drivers: sensor: ina230: driver refactoring

Create an explicit ina230 driver which is supposed to
work with 230 and 231 variants. While at it switch
to i2c_dt_spec helpers and change device-tree node
names to use - instead of _ in order to follow
convention.

Signed-off-by: Bartosz Bilas <b.bilas@grinn-global.com>

bbilas committed:
drivers: sensor: ina23x: add support for INA237

Add the new driver for INA237 variant which is
quite similar to the already supported INA230.

Signed-off-by: Bartosz Bilas <b.bilas@grinn-global.com>

bbilas commented:
@gmarull Drivers have been split into ina230 which handle 230/231 variants and ina237 for 237 variant.

gmarull commented on drivers/sensor/ina23x/Kconfig:
```diff
@@ -10,8 +10,12 @@ config INA23X
-config INA23X_TRIGGER
-	bool "Trigger mode"
+config INA230
+	bool "INA230"
```
missing help

bbilas replied:
Fixed.

gmarull commented on tests/drivers/build_all/sensor/prj.conf:
```diff
@@ -98,4 +98,4 @@ CONFIG_TMP112=y
 CONFIG_VCNL4040=y
 CONFIG_VL53L0X=y
-CONFIG_INA23X=y
+CONFIG_INA230=y
```
will this build without selecting `INA23X`?

bbilas replied:
Right, I missed that. Fixed.

ycsin commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,305 @@
+ * @param current_lsb Value of the Current register LSB in milliamps
+ * @param rshunt Shunt resistor value in milliohms
+ */
+#define INA237_CALIB(current_lsb, rshunt) ((8192 * current_lsb * rshunt) / 100)
```
Is it not necessary to have unsigned suffix for these?

bbilas replied:
I don't see a particular reason to use it here.

mbolivar-nordic commented on dts/bindings/sensor/ti,ina237.yaml:
```diff
@@ -0,0 +1,41 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: |
```
This is duplicating common properties with ti,ina230.yaml. I think this should use a common include file instead. See dts/bindings/serial/nordic,nrf-uart-common.yaml and the places that include it for an example.

bbilas replied:
Fixed.

mbolivar-nordic reviewed (changes requested):
Split driver with shared codebase sounds like a good idea from a driver perspective; let's do some more sharing in the DT too, please.

It's a shame this binding is using underscores for its property names (`current_lsb` instead of `current-lsb`), but oh well, too late...

bbilas commented:
> Split driver with shared codebase sounds like a good idea from a driver perspective; let's do some more sharing in the DT too, please.
> 
> It's a shame this binding is using underscores for its property names (`current_lsb` instead of `current-lsb`), but oh well, too late...

Fixed.

gmarull commented on drivers/sensor/ina23x/ina23x_common.h:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @brief Macro used to test if the current's sign bit is set
+ */
+#define INA23X_CURRENT_SIGN_BIT        0x8000
```
`BIT(15)`?

bbilas replied:
Replaced.

gmarull commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,305 @@
+ * @brief Macros to convert raw Bus voltage to millivolts when current_lsb is
+ *        set to 1mA.
+ *
+ * reg value read from bus voltage register
```
`@param reg...` (and others)

bbilas replied:
Fixed.

gmarull commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,305 @@
+ *        set to 1mA.
+ *
+ * reg value read from power register
+ * clsb value of current_lsb
```
what does this line mean? return value?

bbilas replied:
Nah, wrong copy&paste :) Removed.

gmarull commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,305 @@
+/**
+ * @brief sensor value get
+ *
+ * @return 0 for success
+ * @return -ENOTSUP for unsupported channels
```
use `@retval` for specific values

bbilas replied:
Fixed.

gmarull commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,305 @@
+		return ina23x_reg_write(&config->bus, INA237_REG_CONFIG, data);
+	case SENSOR_ATTR_CALIBRATION:
+		return ina23x_reg_write(&config->bus, INA237_REG_CALIB, data);
+	case SENSOR_ATTR_FEATURE_MASK:
+	case SENSOR_ATTR_ALERT:
```
add `__fallthrough`, or just use `default` (equivalent)

bbilas replied:
Right, fixed.

gmarull commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,305 @@
+			return ret;
+		}
+		break;
+	case SENSOR_ATTR_FEATURE_MASK:
+	case SENSOR_ATTR_ALERT:
+	default:
```
ditto

bbilas replied:
Fixed.

gmarull reviewed (approved):
some minor comments, overall lgtm

bbilas commented:
@mbolivar-nordic gentle ping ;)

mniestroj commented on dts/bindings/sensor/ti,ina230.yaml:
```diff
@@ -1,49 +1,29 @@
       description: Mask register, default matches the power-on reset value
 
-    alert_limit:
+    alert-limit:
```
It is worth noting in the commit message, that various function doxygen documentation has been updated and also some device-tree node names now use `-` instead of `_` in order to follow convention.

bbilas replied:
Fixed.

mniestroj commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,300 @@
+#define INA237_CALIB(current_lsb, rshunt) ((8192 * current_lsb * rshunt) / 100)
+
+/**
+ * @brief Macros to convert raw Bus voltage to millivolts when current_lsb is
```
```suggestion
 * @brief Macro to convert raw Bus voltage to millivolts when current_lsb is
```

bbilas replied:
Fixed.

mniestroj commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,300 @@
+#define INA237_BUS_MV(reg) ((reg * 3125) / 1000)
+
+/**
+ * @brief Macros to convert raw power value to milliwatts when current_lsb is
```
```suggestion
 * @brief Macro to convert raw power value to milliwatts when current_lsb is
```

bbilas replied:
Fixed.

mniestroj commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -0,0 +1,300 @@
+ * @brief Macros to convert raw Bus voltage to millivolts when current_lsb is
+ *        set to 1mA.
+ *
+ * @param value read from bus voltage register.
```
```suggestion
 * @param reg value read from bus voltage register.
```

bbilas replied:
Fixed.

mbolivar-nordic reviewed (changes requested):
>>  It's a shame this binding is using underscores for its property names (current_lsb instead of current-lsb), but oh well, too late...
>
> Fixed.

Thanks for that, but what I meant by "oh well, too late" was that we can't change this property name without breaking people's devicetree files, but I guess this is breaking compatibility anyway since the old compatible is gone.

Can you please add this to the release notes as well?

Something like this:

```patch
diff --git a/doc/releases/release-notes-3.0.rst b/doc/releases/release-notes-3.0.rst
index 83b07d295e..c893d9a35d 100644
--- a/doc/releases/release-notes-3.0.rst
+++ b/doc/releases/release-notes-3.0.rst
@@ -290,6 +290,9 @@ Build and Infrastructure
 
 * Devicetree
 
+  * Support for the devicetree compatible ``ti,ina23x`` has been removed.
+    Instead, use :dtcompatible:`ti,ina230` or :dtcompatible:`ti,ina237`.
+
 
 * West (extensions)
```

With that, I am good with this PR.

MaureenHelm commented on dts/bindings/sensor/ti,ina23x-common.yaml:
```diff
@@ -0,0 +1,28 @@
+    config:
+      type: int
+      required: true
+      description: Configuration register
```
Does this correspond to a hardware register? Please document more clearly and consider setting a default value

bbilas replied:
I've updated the description but I didn't set the default value since this register is different for 230/231 and 237 variants.

MaureenHelm replied:
Ok, then please mention this in this the description.

bbilas replied:
Done.

MaureenHelm commented on drivers/sensor/ina23x/ina230.c:
```diff
@@ -0,0 +1,334 @@
+ * @param current_lsb Value of the Current register LSB in milliamps.
+ * @param rshunt Shunt resistor value in milliohms.
+ */
+ #define INA230_CALIB(current_lsb, rshunt) (5120 / ((current_lsb) * (rshunt)))
+
+/**
+ * @brief Macro to convert raw Bus voltage to millivolts when current_lsb is
+ *        set to 1mA.
+ *
+ * @param reg value read from bus voltage register.
+ */
+#define INA230_BUS_MV(reg) ((reg) * 125 / 100)
+
+/**
+ * @brief Macro to convert raw power value to milliwatts when current_lsb is
+ *        set to 1mA.
+ *
+ * @param reg value read from power register.
+ */
+#define INA230_POW_MW(reg) ((reg) * 25)
```
I'd prefer not to hide these arithmetic operations behind macros, especially since they're only being used once

bbilas replied:
In my opinion it's better to have it as a macro with an explanation since it's more clearer for the user and they are being used twice per channel get request.

MaureenHelm replied:
`INA230_CALIB` is only used once.

`INA230_BUS_MV` and `INA230_POW_MW` are each used twice, but within the same function scope. It would be more clear to use an intermediate local variable with the explanation there, and it will also be easier to see if any necessary typecasting is handled correctly.

bbilas replied:
Macros replaced.

bbilas committed:
CODEOWNERS: Add bbilas for INA23X driver

Add myself as codeowners for INA23X driver files.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
doc: releases: add a note about ina23x driver changes

Let the people know that the property names have been changed
so compatibility will be broken in the v3.0 release.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas commented:
> > > It's a shame this binding is using underscores for its property names (current_lsb instead of current-lsb), but oh well, too late...
> > 
> > 
> > Fixed.
> 
> Thanks for that, but what I meant by "oh well, too late" was that we can't change this property name without breaking people's devicetree files, but I guess this is breaking compatibility anyway since the old compatible is gone.
> 
> Can you please add this to the release notes as well?
> 
> Something like this:
> 
> ```diff
> diff --git a/doc/releases/release-notes-3.0.rst b/doc/releases/release-notes-3.0.rst
> index 83b07d295e..c893d9a35d 100644
> --- a/doc/releases/release-notes-3.0.rst
> +++ b/doc/releases/release-notes-3.0.rst
> @@ -290,6 +290,9 @@ Build and Infrastructure
>  
>  * Devicetree
>  
> +  * Support for the devicetree compatible ``ti,ina23x`` has been removed.
> +    Instead, use :dtcompatible:`ti,ina230` or :dtcompatible:`ti,ina237`.
> +
>  
>  * West (extensions)
> ```
> 
> With that, I am good with this PR.

Note added.

bbilas commented on drivers/sensor/ina23x/ina23x_common.c:
```diff
@@ -0,0 +1,48 @@
+	uint8_t data[2];
+	int ret;
+
+	ret = i2c_burst_read_dt(bus, reg, data, 2);
```
Fixed.

bbilas commented on drivers/sensor/ina23x/ina23x_common.c:
```diff
@@ -0,0 +1,48 @@
+	uint8_t data[3];
+	int ret;
+
+	ret = i2c_burst_read_dt(bus, reg, data, 3);
```
Fixed.

bbilas commented on CODEOWNERS:
```diff
@@ -319,6 +319,7 @@
 /drivers/sensor/ams_iAQcore/              @alexanderwachter
 /drivers/sensor/ens210/                   @alexanderwachter
 /drivers/sensor/hts*/                     @avisconti
+/drivers/sensor/ina23x/ina23x_common*     @bbilas
```
The reason why I've done it in that way was that I didn't want to take ownership of the existing ina230 driver which was created by @sambhurst before but ok, I've changed as you said.

bbilas commented on CODEOWNERS:
```diff
@@ -472,6 +473,7 @@
 /dts/bindings/ethernet/*gem.yaml          @ibirnbaum
 /dts/posix/                               @aescolar @vanwinkeljan @daor-oti
 /dts/bindings/sensor/*bme680*             @BoschSensortec
+/dts/bindings/sensor/*ina23x-common.yaml  @bbilas
```
Fixed.

bbilas commented:
Added missing helpers for dt-bindings.

mbolivar-nordic reviewed (approved):
review is for the DT portions

bbilas commented:
Please set DNM label because I'm not able to verify the latest changes due to lack of hardware. It will be verified by my colleague this week.

bbilas commented:
@MaureenHelm changes have been tested using INA237 and everything looks good

<pre>
uart:~$ sensor get INA237
channel idx=30 voltage =   5.387000
channel idx=31 current =   0.497200
channel idx=32 power =   2.678200
</pre>
